lib/src/rules/null_safety.dart

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2021, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert';
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Do not opt out of null safety.';
+
+const _details = r'''
+
+**DON'T** use unsound null safety, by specifying a dart version lower than `2.12`.
+
+**BAD:**
+```dart
+// @dart=2.8
+a {
+}
+```
+
+**GOOD:**
+```dart
+b {
+}
+''';
+
+class NullSafety extends LintRule implements NodeLintRule {
+  NullSafety()
+      : super(
+            name: 'null_safety',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this, context);
+    registry.addCompilationUnit(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+  final LinterContext context;
+
+  _Visitor(this.rule, this.context);
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    var content = context.currentUnit.content;
+    var lines = LineSplitter().convert(content);
+
+    var regExp = RegExp(r"//@dart\=2\.(\d)");
```

Comment:
This regex doesn't work with all of the possibilities. Users are allowed to have whitespace almost anywhere in the line, and we probably need to catch comments that use a major version of `1` (though that might be disallowed in general; I cant remember). We should probably share the regex used by `LanguageVersionOverrideVerifier` (in the analyzer package). Unfortunately that class isn't public API (and the field is private), so you'll have to copy it with a note that it needs to be kept in sync with the original (though I can't imagine that the format will change).

The use of `firstMatch` below means that it can match even if there's non-whitespace before the comment:
```
void main() { //@dart=2.2
}
```
but that isn't considered a version comment because the version comment has to come before any non-comment in the file. The use of `firstMatch` also means that it could match text in a string literal. Both of those are unlikely possibilities, but there is a slightly different approach that would solve those cases and be more performant when dealing with larger files.

The AST structure provides access to the underlying tokens, including comments. In particular you can ask any node for the comment tokens that proceed it. The class `LanguageVersionOverrideVerifier` uses this to limit the search to only those comments that appear before the first non-comment token. Consider using a similar approach.

Reply:
I see, in the previous code there was `.replaceAll(' ', '')` to remove all whitespaces before using the `RegExp`.

Anyhow, it is now changed to the regular expression from `LanguageVersionOverrideVerifier` which contains the initial line `^`, I believe that would make things better, since it will always consider line beginning.

Also, the comments of the first token are iterated.

Regarding the `PubspecVisitor`, by naked eye, I couldn't see something related to the `environment` or `sdk`, and since this is in `analyzer`, how about restricting this PR scope to the dart files. This `PubspecVisitor` seems to have dependency elsewhere.

Reply:
I missed the line to remove spaces, thanks.

Yes, I'm totally fine with merging this as-is and adding the pubspec handling later. I don't see support for the sdk version either, so we'll need to add that.